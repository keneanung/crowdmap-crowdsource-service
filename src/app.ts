import cors from "cors";
import express, {
  Request as ExRequest,
  Response as ExResponse,
  NextFunction,
  json,
  urlencoded,
} from "express";
import rateLimit from "express-rate-limit";
import path from "node:path";
import swaggerUi from "swagger-ui-express";
import { ValidateError } from "tsoa";
import { RegisterRoutes } from "../generated/routes";
import * as swaggerJson from "../generated/swagger.json";
import { config } from "./config/values";
import {
  AuthorizationError,
  ConflictError,
  NotFoundError,
} from "./models/api/error";

export const app = express();

app.set("trust proxy", config.trustProxy);

app.use(
  urlencoded({
    extended: true,
  }),
);
app.use(json());
app.use(
  rateLimit({
    // use a 15 minute window
    windowMs: 15 * 60 * 1000,
    // allow ten requests per second
    max: 15 * 60 * 10,
    standardHeaders: true,
  }),
);
app.use("/docs", swaggerUi.serve, (_req: ExRequest, res: ExResponse) => {
  return res.send(swaggerUi.generateHTML(swaggerJson));
});

RegisterRoutes(app);

app.use(express.static(path.join(path.dirname(__dirname), "website")));

app.use(function notFoundHandler(_req, res: ExResponse) {
  res.status(404).send({
    message: "Not Found",
  });
});

app.use(function errorHandler(
  err: unknown,
  req: ExRequest,
  res: ExResponse,
  next: NextFunction,
): ExResponse | undefined {
  if (err instanceof ValidateError) {
    console.warn(`Caught Validation Error for ${req.path}:`, err.fields);
    return res.status(422).json({
      message: "Validation Failed",
      details: err.fields,
    });
  }
  if (err instanceof AuthorizationError) {
    console.warn(`Caught Authentication Error for ${req.path}:`, err.message);
    return res.status(403).json({
      message: err.message,
    });
  }
  if (err instanceof ConflictError) {
    console.error(err);
    return res.status(409).json({
      message: err.message,
    });
  }
  if (err instanceof NotFoundError) {
    console.error(err);
    return res.status(404).json({
      message: err.message,
    });
  }
  if (err instanceof Error) {
    console.error(err);
    return res.status(500).json({
      message: "Internal Server Error",
    });
  }

  next();
  return;
});
app.use(cors());
