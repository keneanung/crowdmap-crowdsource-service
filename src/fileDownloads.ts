import * as fs from "fs";
import { Readable } from "stream";
import { config } from "./config/values";

export const downloadMapVersion = async () => {
  try {
    await downloadFile(config.versionDownloadUrl, config.versionFile);
  } catch (err) {
    throw Error("Failed to download version file", {
      cause: err,
    });
  }
};

export const downloadMapFile = async () => {
  try {
    await downloadFile(config.mapDownloadUrl, config.mapFile);
  } catch (err) {
    throw Error("Failed to download map file", {
      cause: err,
    });
  }
};

const downloadFile = async (source: string, destination: string) => {
  const res = await fetch(source);

  const fileStream = fs.createWriteStream(destination);
  if (res.body === null) {
    throw Error("No body");
  }
  const readable = Readable.fromWeb(res.body);
  readable.on("error", (err) => {
    console.error(err);
    throw Error("Failed to download file", {
      cause: err,
    });
  });
  readable.pipe(fileStream);
};
