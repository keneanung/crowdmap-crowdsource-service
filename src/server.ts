import * as fs from "fs";
import { app } from "./app";
import { config } from "./config/values";
import { downloadMapFile, downloadMapVersion } from "./fileDownloads";
import { iocContainer } from "./ioc/ioc";
import { UserService } from "./services/userService";

let mapDownloadPromise;
if (!fs.existsSync(config.mapFile)) {
  mapDownloadPromise = downloadMapFile().catch((err) => {
    console.error(err);
    process.exit(1);
  });
} else {
  mapDownloadPromise = Promise.resolve();
}

let mapVersionDownloadPromise;
if (!fs.existsSync(config.versionFile)) {
  mapVersionDownloadPromise = downloadMapVersion().catch((err) => {
    console.error(err);
    process.exit(1);
  });
} else {
  mapVersionDownloadPromise = Promise.resolve();
}

const userService = iocContainer.resolve<UserService>(UserService);
const checkAdminUser = userService.getUsers().then((users) => {
  const adminUser = users.find((user) => user.name === "admin");
  if (!adminUser) {
    const apiKey = userService.generateApiKey();
    console.log(`Generated admin API Key: ${apiKey}`);
    return userService.addUser("admin", apiKey, ["site_admin", "map_admin"]);
  }
  return Promise.resolve();
});

Promise.all([
  mapDownloadPromise,
  mapVersionDownloadPromise,
  checkAdminUser,
]).then(
  () => {
    app.listen(config.port, () => {
      console.log(
        `Crowdmap service listening at http://localhost:${config.port}`,
      );
    });
  },
  (err) => {
    console.error(err);
    process.exit(1);
  },
);
