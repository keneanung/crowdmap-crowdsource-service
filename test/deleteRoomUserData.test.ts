/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { beforeEach, expect, test } from "@jest/globals";
import request from "supertest";
import { app } from "../src/app";
import { setupChangeServiceMock } from "./setup/iocSetup";

beforeEach(() => {
  setupChangeServiceMock();
});

test("Should accept and return room user data deletions", async () => {
  await request(app)
    .post("/change")
    .send({
      type: "delete-room-user-data",
      roomNumber: 1,
      key: "test user data",
      reporter: "Test Reporter",
    })
    .expect(201);

  await request(app)
    .get("/change")
    .expect(200)
    .expect((res) => {
      expect(res.body).toStrictEqual([
        {
          type: "delete-room-user-data",
          roomNumber: 1,
          key: "test user data",
          reporters: 1,
        },
      ]);
    });
});

test("Should incorporate room user data deletions into the map", async () => {
  await request(app).post("/change").send({
    type: "delete-room-user-data",
    roomNumber: 1,
    key: "gotoMapping",
    reporter: "Test Reporter",
  });

  await request(app)
    .get("/map?format=json&timesSeen=0")
    .expect(200)
    .expect("X-Map-Version", "466.1.1")
    .expect((res) => {
      const responseText = res.text;
      const map: any = JSON.parse(responseText);
      const area: any = map.areas[5];
      const room: any = area.rooms[0];
      expect(room).toMatchSnapshot();
    });
});
