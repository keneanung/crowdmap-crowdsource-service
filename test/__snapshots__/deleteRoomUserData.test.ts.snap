// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should incorporate room user data deletions into the map 1`] = `
{
  "coordinates": [
    32568,
    0,
    0,
  ],
  "environment": -1,
  "exits": [],
  "id": 1,
  "userData": {
    "Game Area": "Haila-ele",
    "areaContinents": "{"Outer":[192,193,162,194,206,351,209,294,180,297,181,366,355,207,191],"Western_Isles":[162,209,207],"Eastern_Isles":[192,180,181,191],"Arcadia":[150,154],"North":[385,261,197,99,199,34,200,101,202,204,352,353,73,242,147,243,74,212,244,213,214,246,301,108,306,383,308,307],"Northern_Isles":[194,206,351,294],"Island":[224,161,225,195,98,164,165,464,240,91,317,92,314],"Main":[127,158,316,2,3,4,5,7,10,12,14,16,20,24,28,32,40,48,56,64,80,96,112,128,160,320,384,448,386,65,81,387,113,130,388,452,389,131,326,33,41,49,57,66,82,196,264,265,133,394,83,115,134,166,198,268,269,167,17,21,25,29,42,50,58,68,84,116,136,272,400,273,401,137,201,274,402,69,85,117,138,139,203,235,278,470,43,51,59,279,102,118,280,472,281,173,205,346,71,87,103,119,142,174,238,284,348,349,477,9,11,13,22,26,30,36,44,60,72,104,120,144,208,145,290,89,105,121,178,293,357,179,211,37,53,61,90,106,122,148,296,471,463,361,425,149,298,362,426,263,75,107,123,182,300,364,291,365,295,151,302,23,27,31,38,46,54,62,76,124,152,398,368,185,391,77,125,186,267,372,318,367,309,359,155,187,350,310,374,55,63,78,375,110,126,156,319,376,315,270,311,303,299,157,275,259,175,79,95,111,159,462,379,465,443,380,444,378,381,390,377,393,35,489,468,476,466,473,474,488,450,475,467,490,491,109],"Meropis":[256,322,325,328,329,330,67,332,396,333,399,336,337,338,340,341,342,343,344,345,141,282,285,286,210,487,445,339,215,305,217,218,334,347,335,39,331,327,283,492,494],"Arena":[228,229,230,231,232,233,369],"Off":[190,380,444,6,257,321,449,324,453,163,390,454,455,456,393,457,458,460,397,168,465,169,170,404,405,469,171,406,35,408,409,410,412,413,15,176,288,416,480,289,417,481,418,482,292,356,420,484,421,485,358,422,45,483,479,478,462,459,245,451,446,443,428,439,363,429,183,379,184,248,415,432,431,430,427,423,419,433,414,411,249,407,434,395,382,381,371,435,436,373,437,438,47,252,440,377,441,271,253,378,442,255,93,100]}",
    "audit.remapped_area": "0",
    "ferry rooms": "[31545,6558,6559,31520,11654,31562,6550,6551,18439,38156,39208,19432,31605]",
  },
}
`;
